rust/ql/lib/codeql/rust/security/UncontrolledAllocationSizeExtensions.qll

```diff
@@ -0,0 +1,95 @@
+/**
+ * Provides classes and predicates for reasoning about uncontrolled allocation
+ * size vulnerabilities.
+ */
+
+import rust
+private import codeql.rust.Concepts
+private import codeql.rust.dataflow.DataFlow
+private import codeql.rust.dataflow.FlowSink
+private import codeql.rust.controlflow.ControlFlowGraph as Cfg
+private import codeql.rust.controlflow.CfgNodes as CfgNodes
+
+/**
+ * Provides default sources, sinks and barriers for detecting uncontrolled
+ * allocation size vulnerabilities, as well as extension points for adding your own.
+ */
+module UncontrolledAllocationSize {
+  /**
+   * A data flow sink for uncontrolled allocation size vulnerabilities.
+   */
+  abstract class Sink extends QuerySink::Range {
+    override string getSinkType() { result = "UncontrolledAllocationSize" }
+  }
+
+  /**
+   * A barrier for uncontrolled allocation size vulnerabilities.
+   */
+  abstract class Barrier extends DataFlow::Node { }
+
+  /**
+   * A sink for uncontrolled allocation size from model data.
+   */
+  private class ModelsAsDataSink extends Sink {
+    ModelsAsDataSink() { sinkNode(this, ["alloc-size", "alloc-layout"]) }
+  }
+
+  /**
+   * A barrier for uncontrolled allocation size that is an upper bound check / guard.
+   */
+  private class UpperBoundCheckBarrier extends Barrier {
```

Comment:
Is it correct that if we had a guards library for Rust, then we could've used that here instead?

Reply:
This uses the barrier guard functionality in `DataFlow.qll` that was added [here](https://github.com/github/codeql/pull/18960/files#diff-141cae14807fa6bc6cfa1a952590c8d0f7845e778570fb91ffc19c00676fd8d1).  It seems to be sufficient.

Which guards library are you thinking of?

I wonder if you're thinking of range analysis actually - that would provide an actual *value* for the upper bound, which we don't really need here, but it would probably also improve accuracy just by covering less common cases better (like `v < 2 * v` in the test).